### PR TITLE
NH-4063 - Fix ODBC failing tests

### PR DESCRIPTION
**Pull request by fredericDelaporte**
[NH-4063](https://nhibernate.jira.com/browse/NH-4063) - Fix ODBC failing tests

### CODE REVIEW

File: src/NHibernate.Test/Hql/HQLFunctions.cs [hunk 1]
```diff
@@ -762,7 +765,7 @@ public void Cast()
 							// This test raises an exception in SQL Server because named 
 							// parameters internally are always positional (@p0, @p1, etc.)
 							// and named differently hence they mismatch between GROUP BY and HAVING clauses.
-							if (!ex.InnerException.Message.Equals(msgToCheck))
+							if (!ex.InnerException.Message.Contains(msgToCheck))
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/Hql/HQLFunctions.cs]: ODBC prepends the message with some additional information.

File: src/NHibernate.Test/Insertordering/InsertOrderingFixture.cs [hunk 2]
```diff
@@ -33,7 +34,7 @@ protected override string MappingsAssembly
 		protected override bool AppliesTo(Dialect.Dialect dialect)
 		{
 			// Custom batcher enforces sql server.
-			return dialect.SupportsSqlBatches;
+			return dialect is MsSql2000Dialect;
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/Insertordering/InsertOrderingFixture.cs]: `SupportsSqlBatches` is only about the `GO` statement handled by SQL Server Client tools. All its usages in tests were invalids.

File: src/NHibernate.Test/NHSpecificTest/NH1756/Fixture.cs [hunk 3]
```diff
@@ -12,6 +14,25 @@ protected override bool AppliesTo(Dialect.Dialect dialect)
 			return dialect is MsSql2000Dialect;
 		}
 
+		protected override bool AppliesTo(ISessionFactoryImplementor factory)
+		{
+			// ODBC driver DateTime handling with SQL Server 2008+ Client is broken and forbids using it as a time stamp
+			// generated on db-side.
+			// Due to NH-3895, we have to force the scale on date-time parameters to 3 (3 significant fractional seconds digits)
+			// when using ODBC + SQL Server 2008+, otherwise DateTime values having milliseconds will be rejected. But the SQL
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/NH1756/Fixture.cs]: A bug, but an external issue against which NHibernate could nothing IMO.

File: src/NHibernate.Test/NHSpecificTest/NH2057/Fixture.cs [hunk 4]
```diff
@@ -1,31 +1,30 @@
...
 			using (var tx = new TransactionScope())
 			{
-				using (s = (SessionImpl)OpenSession())
+				using (s = OpenSession())
 				{
 					s.Get<Person>(1);
 				}
 				//not closed because the tx is opened yet
-				Assert.False(s.IsClosed);
+				Assert.False(s.GetSessionImplementation().IsClosed);
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/NH2057/Fixture.cs]: There is `ISession.IsOpen` please check it instead

File: src/NHibernate.Test/NHSpecificTest/NH2057/Fixture.cs [hunk 5]
```diff
@@ -1,31 +1,30 @@
...
 				tx.Complete();
 			}
-			Assert.That(s.IsClosed, Is.True);
+			// ODBC does promote to distributed, causing the completion to happen concurrently to code
+			// following scope disposal. Sleep a bit for accounting for this.
+			Thread.Sleep(100);
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/NH2057/Fixture.cs]: NUnit provides a way to check something after some time. Please use it.

**Pull request merged**